esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5000,12 +5000,75 @@ bool CWsDfuEx::onDFUBrowseData(IEspContext &context, IEspDFUBrowseDataRequest &r
         }
     }
     catch(IException* e)
-    {   
+    {
+        FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+
+bool CWsDfuEx::onListHistory(IEspContext &context, IEspListHistoryRequest &req, IEspListHistoryResponse &resp)
+{
+    try
+    {
+        StringBuffer username;
+        context.getUserID(username);
+        Owned<IUserDescriptor> userdesc;
+        if(username.length() > 0)
+        {
+            const char* passwd = context.queryPassword();
+            userdesc.setown(createUserDescriptor());
+            userdesc->set(username.str(), passwd);
+        }
+
+        if (!req.getName() || !*req.getName())
+            throw MakeStringException(ECLWATCH_MISSING_PARAMS, "Name required");
+        PROGLOG("onListHistory: %s", req.getName());
+
+        Owned<IDFUhelper> dfuhelper = createIDFUhelper();
+        MemoryBuffer xmlmap;
+        dfuhelper->getFileHistory(req.getName(), xmlmap, userdesc.get());
+        resp.setXmlmap(xmlmap);
+    }
+    catch(IException* e)
+    {
         FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
     }
     return true;
 }
 
+bool CWsDfuEx::onEraseHistory(IEspContext &context, IEspEraseHistoryRequest &req, IEspEraseHistoryResponse &resp)
+{
+    try
+    {
+        StringBuffer username;
+        context.getUserID(username);
+        Owned<IUserDescriptor> userdesc;
+        if(username.length() > 0)
+        {
+            const char* passwd = context.queryPassword();
+            userdesc.setown(createUserDescriptor());
+            userdesc->set(username.str(), passwd);
+        }
+
+        if (!req.getName() || !*req.getName())
+            throw MakeStringException(ECLWATCH_MISSING_PARAMS, "Name required");
+        PROGLOG("onEraseHistory: %s", req.getName());
+
+        Owned<IDFUhelper> dfuhelper = createIDFUhelper();
+        MemoryBuffer xmlmap;
+        dfuhelper->getFileHistory(req.getName(), xmlmap, userdesc.get());
+        resp.setXmlmap(xmlmap);
+
+        Owned<IDistributedFile> file = queryDistributedFileDirectory().lookup(req.getName(),userdesc.get());
```

Comment:
Should get the history here (after checking file exists) and then serialize it, instead of lines above.